system/security/shared/caching.cpp

```diff
@@ -672,16 +681,39 @@ bool CPermissionsCache::queryPermsManagedFileScope(ISecUser& sec_user, const cha
 
 SecAccessFlags CPermissionsCache::queryDefaultPermission(ISecUser& user)
 {
-    if (m_defaultPermission == SecAccess_Unknown)
+    if (!m_secMgr)
+        return SecAccess_Full; // if no security manager, all full access to all scopes
+
+    if (m_useLegacyDefaultFileScopePermissionCache)
     {
-        if (m_secMgr)
+        if (m_defaultPermission == SecAccess_Unknown)
+        {
             m_defaultPermission = m_secMgr->queryDefaultPermission(user);
-        else
-            m_defaultPermission = SecAccess_None;
+            DBGLOG("Legacy default file scope permission set to %s(%d) for all users, based on User '%s'", getSecAccessFlagName(m_defaultPermission),
+                   m_defaultPermission, user.getName());
+        }
+        return m_defaultPermission;
     }
-    return m_defaultPermission;
 
+    SecAccessFlags defaultPermission = SecAccess_None;
+    CriticalBlock defaultScopePermissionBlock(syncDefaultScopePermissions);
+    const std::string username(user.getName());
+    auto it = m_userDefaultFileScopePermissions.find(username);
+    if (it == m_userDefaultFileScopePermissions.end())
+    {
+        defaultPermission = m_secMgr->queryDefaultPermission(user);
+        m_userDefaultFileScopePermissions.emplace(username, defaultPermission);
+        DBGLOG("Added user '%s' to default file scope permissions with access %s(%d)", username.c_str(), getSecAccessFlagName(defaultPermission),
```

Comment:
It is best to avoid any logging inside a critical section - because it could be blocked on io.  Better to set a flag and trace once the critical section is left.